File: npc_sim/core/sim_vector3.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Unity-free 3D vector for use throughout the simulation core."""

from __future__ import annotations
import math
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class SimVector3:
    """Immutable 3D vector. All simulation positions and directions use this type."""
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0

    # ── Class-level constants (set after class definition) ──

    @property
    def sqr_magnitude(self) -> float:
        return self.x * self.x + self.y * self.y + self.z * self.z

    @property
    def magnitude(self) -> float:
        return math.sqrt(self.sqr_magnitude)

    def normalized(self) -> SimVector3:
        m = self.magnitude
        if m < 1e-6:
            return SimVector3.ZERO
        return SimVector3(self.x / m, self.y / m, self.z / m)

    def within_radius(self, other: SimVector3, radius: float) -> bool:
        return SimVector3.sqr_distance(self, other) <= radius * radius

    # ── Static math ──

    @staticmethod
    def sqr_distance(a: SimVector3, b: SimVector3) -> float:
        dx = a.x - b.x
        dy = a.y - b.y
        dz = a.z - b.z
        return dx * dx + dy * dy + dz * dz

    @staticmethod
    def distance(a: SimVector3, b: SimVector3) -> float:
        return math.sqrt(SimVector3.sqr_distance(a, b))

    # ── Operators ──

    def __add__(self, other: SimVector3) -> SimVector3:
        return SimVector3(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other: SimVector3) -> SimVector3:
        return SimVector3(self.x - other.x, self.y - other.y, self.z - other.z)

    def __mul__(self, scalar: float) -> SimVector3:
        return SimVector3(self.x * scalar, self.y * scalar, self.z * scalar)

    def __rmul__(self, scalar: float) -> SimVector3:
        return self.__mul__(scalar)

    def __repr__(self) -> str:
        return f"({self.x:.2f}, {self.y:.2f}, {self.z:.2f})"


# Class-level constants
SimVector3.ZERO = SimVector3(0.0, 0.0, 0.0)
SimVector3.ONE = SimVector3(1.0, 1.0, 1.0)

File: Runtime/Core/SimVector3.cs
using System;

namespace ForgeProject.Sim
{
    /// <summary>
    /// Unity-free 3D vector for use throughout the simulation core.
    /// Keeps all domain logic independent of UnityEngine.
    /// </summary>
    [Serializable]
    public readonly struct SimVector3 : IEquatable<SimVector3>
    {
        public readonly float X;
        public readonly float Y;
        public readonly float Z;

        public static readonly SimVector3 Zero = new(0f, 0f, 0f);
        public static readonly SimVector3 One  = new(1f, 1f, 1f);

        public SimVector3(float x, float y, float z)
        {
            X = x; Y = y; Z = z;
        }

        // ─── Math ─────────────────────────────────────────────────────────────────

        public float SqrMagnitude => X * X + Y * Y + Z * Z;
        public float Magnitude    => MathF.Sqrt(SqrMagnitude);

        /// <summary>Euclidean distance squared — avoids sqrt for range checks.</summary>
        public static float SqrDistance(SimVector3 a, SimVector3 b)
        {
            float dx = a.X - b.X;
            float dy = a.Y - b.Y;
            float dz = a.Z - b.Z;
            return dx * dx + dy * dy + dz * dz;
        }

        /// <summary>Euclidean distance.</summary>
        public static float Distance(SimVector3 a, SimVector3 b)
            => MathF.Sqrt(SqrDistance(a, b));

        /// <summary>Returns true when <paramref name="other"/> is within <paramref name="radius"/>.</summary>
        public bool WithinRadius(SimVector3 other, float radius)
            => SqrDistance(this, other) <= radius * radius;

        public SimVector3 Normalized()
        {
            float m = Magnitude;
            return m > 1e-6f ? new SimVector3(X / m, Y / m, Z / m) : Zero;
        }

        // ─── Operators ────────────────────────────────────────────────────────────

        public static SimVector3 operator +(SimVector3 a, SimVector3 b) => new(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
        public static SimVector3 operator -(SimVector3 a, SimVector3 b) => new(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
        public static SimVector3 operator *(SimVector3 v, float s)      => new(v.X * s, v.Y * s, v.Z * s);
        public static SimVector3 operator *(float s, SimVector3 v)      => v * s;

        // ─── Equality ─────────────────────────────────────────────────────────────

        public bool Equals(SimVector3 other) => X == other.X && Y == other.Y && Z == other.Z;
        public override bool Equals(object obj) => obj is SimVector3 v && Equals(v);
        public override int GetHashCode() => HashCode.Combine(X, Y, Z);
        public static bool operator ==(SimVector3 a, SimVector3 b) => a.Equals(b);
        public static bool operator !=(SimVector3 a, SimVector3 b) => !a.Equals(b);

        public override string ToString() => $"({X:F2}, {Y:F2}, {Z:F2})";
    }
}
